### PR TITLE
Get local mode working 1/n

### PR DESCRIPTION
**Pull request by eric-anderson**
Working for a few unit tests (file writer), examples (html_ingest, ndd_debug), and notebooks (default-prep-script).
* Update the file writers to work in local mode
* Refactor the code for the database writers to remove duplicate lines.
* Refactor infer_fs so we can use it in multiple places and a variant that will cross check that the user specified fs and the inferred one are the same, in which case we use the inferred path to be more similar to ray.
* Add constants for exec mode to sycamore namespace. Remove materialize constants from default imports.
* Add materialize in a bunch of places to speed up execution.
* Discover we have a strong ray dependency for specifing parallelism. Add TODO to fix this problem.

### CODE REVIEW

File: lib/sycamore/sycamore/writer.py [hunk 1]
```diff
@@ -742,6 +711,13 @@ def json(
...
+
+        self._maybe_execute(node, True)
+
+    def _maybe_execute(self, node: Node, execute: bool) -> Optional[DocSet]:
+        ds = DocSet(self.context, node)
+        if not execute:
+            return ds
+
+        ds.execute()
+        return None
```
Comment (HenryL27) [inline lib/sycamore/sycamore/writer.py:723]: nice

File: notebooks/default-prep-script.ipynb [hunk 2]
```diff
@@ -312,7 +312,16 @@
...
    "execution_count": null,
    "id": "8",
-   "metadata": {},
+   "metadata": {
+    "nbmake": {
+     "post_cell_execute": [
+      "if sycamore.EXEC_LOCAL != sycamore.ExecMode.RAY:",
+      "    import sys",
+      "    # TODO: eric - remove the True",
+      "    assert True or 'ray' not in sys.modules"
```
Comment (HenryL27) [inline notebooks/default-prep-script.ipynb:321]: what does this do?

**Review by HenryL27 — APPROVED** [review]
lgtm. A little confused on what dropping those constants from sycamore __ all__ did.

File: lib/sycamore/sycamore/__init__.py [hunk 3]
```diff
@@ -14,6 +16,4 @@
     "Execution",
     "ExecMode",
     "MaterializeSourceMode",
-    "MATERIALIZE_RECOMPUTE",
-    "MATERIALIZE_USE_STORED",
```
[the remark this answers is not in the document]
Comment (eric-anderson) [reply lib/sycamore/sycamore/__init__.py:18]: It works fine. __all__ affects what happens if you do import sycamore.*

File: lib/sycamore/sycamore/materialize.py [hunk 4]
```diff
@@ -224,22 +224,11 @@ def finalize(self):
...
-            # normalize everything to a URI.
-            p = Path(path)
-            if p.is_absolute():
-                path = p.as_uri()
-            else:
-                path = p.absolute().as_uri()
-
-        from pyarrow import fs
+    def infer_fs(path: str) -> Tuple["pyarrow.FileSystem", Path]:
+        from sycamore.utils.pyarrow import infer_fs
```
[the remark this answers is not in the document]
Comment (eric-anderson) [reply lib/sycamore/sycamore/materialize.py:228]: done.

File: notebooks/default-prep-script.ipynb (hunk 2 above)
Comment (eric-anderson) [reply notebooks/default-prep-script.ipynb:321]: Lets you add "tests" to notebooks. If we're running in local mode, then ray should not be in sys.modules.
I was planning on having the test script run the notebooks and then replace EXEC_LOCAL with EXEC_RAY and then running them a second time to make sure the notebooks work in both modes.  However, the check I wanted to add doesn't work yet.

**Pull request merged**